articles/ros2_security_enclaves.md

```diff
@@ -0,0 +1,315 @@
+---
+layout: default
+title: ROS 2 Security Enclaves
+permalink: articles/ros2_security_enclaves.html
+abstract:
+  This article specifies the integration of security enclaves.
+author:  >
+  [Ruffin White](https://github.com/ruffsl),
+  [Mikael Arguedas](https://github.com/mikaelarguedas)
+
+published: true
+categories: Security
+---
+
+{:toc}
+
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+This design document formalizes the integration of ROS 2 with security enclaves.
+In summary, all processes must load a common enclave that contains the runtime security artifacts unique to that enclave, yet each process may not necessarily have a unique enclave.
+Multiple enclaves can be encapsulated in a single security policy to accurately model the information flow control.
+Users can tune the fidelity of such models by controlling at what scope enclaves are applied at deployment.
+E.g. one unique enclave per OS process, or per OS user, or per device/robot, or per swarm, etc.
+The rest of this document details how enclaves can be organized and used by convention.
+
+## Concepts
+
+Before detailing the SROS 2 integration of the enclaves, the following concepts are introduced.
+
+### Participant
+
+Participant is the object representing a single entity on the network.
+In the case of DDS, the ``Participant`` is a DDS DomainParticipant, which has both access control permissions and a security identity.
+
+### Namespaces
+
+Namespaces are a fundamental design pattern in ROS and are widely used to organize and differentiate many types of resources as to be uniquely identifiable; i.e. for topics, services, actions, and node names.
+As such, the concept of namespacing is well known and understood by current users, as well as strongly supported by the existing tooling.
+Namespaces are often configurable at runtime via command line arguments or statically/programmatically via launch file declarations.
+
+Previously, the Fully Qualified Name (FQN) of a node was used directly by a selected security directory lookup strategy to load the necessary key material.
+However, now that Participants map to contexts and not nodes, such a direct mapping of node FQN to security artifacts is no longer appropriate.
+
+### Contexts
+
+With the advent of ROS 2, multiple nodes may now be composed into one process for improved performance.
+Previously however, each node would retain it's one to one mapping to a separate middleware ``Participant``.
+Given the non-negligible overhead incurred of multiple ``Participant``s per process, a change was introduced to map a single ``Participant`` to a context, and allow for multiple nodes to share that context.
```

Comment:
It would be good to add an explanation of how you can have many contexts (i.e.: many participants), but if all of them are in the same process they will use the same security enclave.

Reply:
Done: d29e06a